eclair-core/src/test/scala/fr/acinq/eclair/wire/OpenTlvSpec.scala

```diff
@@ -16,22 +16,43 @@
 
 package fr.acinq.eclair.wire
 
+import fr.acinq.eclair.UInt64
 import fr.acinq.eclair.channel.ChannelVersion
+import fr.acinq.eclair.wire.ChannelTlv.UpfrontShutdownScript
+import fr.acinq.eclair.wire.CommonCodecs.{varint, varintoverflow}
 import fr.acinq.eclair.wire.OpenChannelTlv.ChannelVersionTlv
+import fr.acinq.eclair.wire.TlvCodecs.tlvStream
 import org.scalatest.funsuite.AnyFunSuite
-import scodec.{Attempt, DecodeResult}
 import scodec.bits._
+import scodec.codecs.{bits, bytes, discriminated, variableSizeBytesLong}
+import scodec.{Attempt, Codec, DecodeResult}
 
 class OpenTlvSpec extends AnyFunSuite {
 
   test("channel version tlv") {
+    case class TestCase(expected: ChannelVersion, encoded: BitVector, reEncoded: BitVector)
+    val testCases = Seq(
+      TestCase(ChannelVersion.STANDARD, hex"fe47000000 00000001".bits, hex"fe47000000 00000001".bits),
+      TestCase(ChannelVersion.STANDARD, hex"fe47000001 04 00000001".bits, hex"fe47000000 00000001".bits),
+      TestCase(ChannelVersion.STANDARD | ChannelVersion.ZERO_RESERVE, hex"fe47000000 00000009".bits, hex"fe47000000 00000009".bits),
+      TestCase(ChannelVersion.STANDARD | ChannelVersion.ZERO_RESERVE, hex"fe47000001 04 00000009".bits, hex"fe47000000 00000009".bits)
+    )
+
+    for (testCase <- testCases) {
+      assert(OpenChannelTlv.openTlvCodec.decode(testCase.encoded) === Attempt.successful(DecodeResult(TlvStream(ChannelVersionTlv(testCase.expected)), BitVector.empty)))
+      assert(OpenChannelTlv.openTlvCodec.encode(TlvStream(ChannelVersionTlv(testCase.expected))) === Attempt.Successful(testCase.reEncoded))
+    }
+  }
 
-    val refs = Map(
-      ChannelVersion.STANDARD -> hex"fe47000000 00000001".bits,
-      (ChannelVersion.STANDARD | ChannelVersion.ZERO_RESERVE) -> hex"fe47000000 00000009".bits
+  test("channel version tlv backwards-compatibility") {
+    // This is the codec that was used previously.
+    val previousCodec: Codec[TlvStream[OpenChannelTlv]] = tlvStream(discriminated[OpenChannelTlv].by(varint)
+      .typecase(UInt64(0), variableSizeBytesLong(varintoverflow, bytes).as[UpfrontShutdownScript])
+      .typecase(UInt64(0x47000000), bits(ChannelVersion.LENGTH_BITS).as[ChannelVersion].as[ChannelVersionTlv])
     )
 
-    refs.foreach { case (cv, bits) => assert(OpenChannelTlv.openTlvCodec.encode(TlvStream(ChannelVersionTlv(cv))) === Attempt.Successful(bits)) }
-    refs.foreach { case (cv, bits) => assert(OpenChannelTlv.openTlvCodec.decode(bits) === Attempt.successful(DecodeResult(TlvStream(ChannelVersionTlv(cv)), BitVector.empty))) }
+    assert(previousCodec.decode(hex"fe47000001 04 00000001".bits).require.value === TlvStream(Nil, Seq(GenericTlv(UInt64(0x47000001), hex"00000001"))))
+    assert(previousCodec.decode(hex"fe47000001 04 00000009".bits).require.value === TlvStream(Nil, Seq(GenericTlv(UInt64(0x47000001), hex"00000009"))))
```

Comment:
The test is okay, I would have sent both tlvs (new and old) to this old codec and make sure it was ignoring the new tlv and interpreting the old one). Seems closer to our use case?

Reply:
The encoder will never add the two in the same tlv stream since it's the same Scala type.
On the contrary, what whill happen is that eclair-node will continue sending the old encoding, but once Phoenix is upgraded to support the new one, we can switch eclair-node to use the new encoding when writing.
Does that make sense?

Reply:
Gotcha. But the test is misleading: even if old Phoenix can parse the new format, they will still break because they will use an invalid channel version, different from the funder. If we want a smooth transition, we need to modify the encoding to have it write both tags.

Reply:
I removed that test. We'll guarantee backwards-compatibility by simply waiting for users to upgrade, as writing both tags will require hacks to the tlv stream codecs...